api/app/weather_models/precip_rdps_model.py

```diff
@@ -0,0 +1,79 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import numpy
+from numba import vectorize
+from app.utils.s3 import read_into_memory
+from app.weather_models import ModelEnum
+from app.weather_models.rdps_filename_marshaller import compose_computed_rdps_filename, compose_rdps_filename
+
+
+@dataclass
+class TemporalPrecip:
+    """Precip values at certain times"""
+
+    timestamp: str
+    precip_amount: float | numpy.ndarray
+
+    def is_after(self, other) -> bool:
+        return self.timestamp > other.timestamp
+
+
+RDPS_S3_PREFIX = f"weather_models/{ModelEnum.RDPS.lower()}/"
+
+
+async def generate_24_hour_accumulating_precip_raster(current_time: datetime):
+    """
+    Given a UTC datetime, grab the raster for that date
+    and the date for 24 hours before to compute the difference.
+    """
+    (yesterday_key, today_key) = get_raster_keys_to_diff(current_time)
+    day_data = await read_into_memory(f"{RDPS_S3_PREFIX}/{today_key}")
+
+    yesterday_time = current_time - timedelta(days=1)
+    yesterday_data = await read_into_memory(f"{RDPS_S3_PREFIX}/{yesterday_key}")
+    if day_data is None or yesterday_data is None:
+        raise ValueError("No precip raster data for %s or %s", today_key, yesterday_key)
+
+    later_precip = TemporalPrecip(timestamp=current_time, precip_amount=day_data)
+    earlier_precip = TemporalPrecip(timestamp=yesterday_time, precip_amount=yesterday_data)
+    return compute_precip_difference(later_precip, earlier_precip)
+
+
+def get_raster_keys_to_diff(timestamp: datetime):
+    """
+    To maintain the invariant that only the same model runs are diffed, we need to grab the latest
+    model run that has data greater than 24 hours ago, but less than 36 hours ago.
+    """
+    target_model_run_date = timestamp - timedelta(hours=24)
+    key_prefix = f"weather_models/{ModelEnum.RDPS.lower()}/{target_model_run_date.date().isoformat()}"
+    # From earlier model run, get the keys for 24 hours before timestamp and the timestamp to perform the diff
+    earlier_key = f"{key_prefix}/"
+    later_key = f"{key_prefix}/"
+    if target_model_run_date.hour != 0 and target_model_run_date.hour != 12:
+        # we're not looking at a run hour, so prefix key with computed path
+        earlier_key = (
+            earlier_key
+            + f"computed/{target_model_run_date.hour:02d}/precip/{compose_computed_rdps_filename(target_model_run_date, target_model_run_date.hour, target_model_run_date.hour)}"
+        )
+        later_key = later_key + f"computed/{timestamp.hour:02d}/precip/{compose_computed_rdps_filename(timestamp, timestamp.hour, timestamp.hour)}"
+    else:
+        earlier_key = earlier_key + f"{target_model_run_date.hour:02d}/precip/{compose_rdps_filename(target_model_run_date, target_model_run_date.hour, target_model_run_date.hour)}"
```

Comment:
There is an edge case we'll need to consider before going live with the calculations. The numerical weather models don't give us a grib file for the first hour of a model run. Precip gribs start at hour 001. So if we're calculating the precip at 0:00 UTC or 12:00 UTC we don't need to do a subtraction (can't actually do the subtraction), the accumulated precip will be equal to the values in the 024 hour grib.